minionsai/discriminator_only/translator.py

```diff
@@ -57,30 +85,122 @@ def translate(self, game: Game):
         
-        return {
-            'board': np.array([board_obs]),  # shape is [batch, num_items=BOARD_SIZE^2]
-            'remaining_turns': np.array([[remaining_turns]]),  # shape is [batch, num_items=1]
-            'money': np.array([[money]]),
-            'opp_money': np.array([[opp_money]]),
-            'score_diff': np.array([[score_diff]])
-        }
+        d = OrderedDict()
+        d['board'] = np.array([board_obs])  # shape is [batch, num_items=BOARD_SIZE^2]
+        d['money'] = np.array([[money]])
+        d['remaining_turns'] = np.array([[remaining_turns]])  # shape is [batch, num_items=1]
+        d['opp_money'] = np.array([[opp_money]])
+        d['score_diff'] = np.array([[score_diff]])
+        if self.mode == "generator":
+            d['phase'] = np.array([[phase]])
+            # legal_actions = [self.valid_actions(game).flatten()]
+            # d['legal_actions'] = np.array([])
+        return d
+
+
+
+    def get_possible_legal_actions(self):
+        possibly_legal = np.zeros((self.num_things, self.num_things), bool)
+        for i in range(BOARD_SIZE ** 2 + 1):
+            for j in range(BOARD_SIZE ** 2 + 1):
+                # money -> board = spawn
+                # money -> money = advance phase
+                if i == self.spawn_pos:
+                    possibly_legal[i,j] = True
+                else:
+                    if j == self.spawn_pos:
+                        continue
+                    x1, y1 = convert_loc_to_pos(i)
+                    x2, y2 = convert_loc_to_pos(j)
+                    if abs(x1-x2) <= MAX_SPEED_OR_RANGE and abs(y1-y2) <= MAX_SPEED_OR_RANGE:
+                        possibly_legal[i,j] = True
+        return possibly_legal
+
+    def valid_actions(self, game: Game):
+        legal = np.zeros((self.num_things, self.num_things), bool)
+        for i in range(self.num_things):
+            for j in range(self.num_things):
+                if self.possibly_legal_actions[i,j]:
+                    if i == self.spawn_pos:
+                        # advance phase action
+                        if j == self.spawn_pos:
+                            legal[i,j] = True
+                            continue
+                        # wrong phase, action is illegal
+                        if game.phase != Phase.SPAWN: 
+                            continue
+                        # spawn action
+                        x2, y2 = convert_loc_to_pos(j)
+                        act = SpawnAction(ZOMBIE, (x2,y2))
+                        if game.process_single_spawn(act, False)[0]:
+                            legal[i,j] = True
+                    else:
+                        # wrong phase, action is illegal
+                        if game.phase != Phase.MOVE:
+                            continue
+                        # move action
+                        x1, y1 = convert_loc_to_pos(i)
```

Comment:
nitpick: can you avoid code duplication by using:
```
act = self.untranslate_action((i, j)
```
instead of duplicating the code here?

In fact would something go wrong if this whole for loop were just:
```
act = self.untranslate_action(i, j)
result, _ = game.process_single_action(act, False)
legal[i, j] = result
```
?

Reply:
the game engine handles trying to move during the wrong phase differently than it handles trying to take an illegal action during a given phase, for some reason. In the former case, the code will crash, while in the latter case, it will return `False` and a status code. I simplified the code a bit, but it'll require a nontrivial refactor of the game engine (and things that relied on the game engine throwing errors instead of getting into weird, unreachable states) to make that work